src/solvers/refinement/string_constraint_generator_transformation.cpp

```diff
@@ -359,38 +360,57 @@ string_exprt string_constraint_generatort::add_axioms_for_char_set(
   return res;
 }
 
+static optionalt<std::pair<exprt, exprt>> try_char_char(
+  exprt needle, exprt replacement)
+{
+  if(needle.type().id()==ID_char && replacement.type().id()==ID_char)
+    return std::make_pair(needle, replacement);
+  const auto needle_str=to_string_expr(needle);
+  const auto replacement_str=to_string_expr(replacement);
+  const auto needle_length=expr_cast<size_t>(needle_str.length());
+  const auto replacement_length=expr_cast<size_t>(replacement_str.length());
+  if(needle_length && replacement_length
+    && *needle_length==1 && *replacement_length==1) {
+    return std::make_pair(exprt(needle_str[0]), exprt(replacement_str[0]));
+  }
+  return { };
+}
+
 /// add axioms corresponding to the String.replace java function
 /// \par parameters: function application with three arguments, the first is a
 ///   string,
-/// the second and the third are characters
+/// the second and the third are either pair of characters or a pair of strings
 /// \return a new string expression
 string_exprt string_constraint_generatort::add_axioms_for_replace(
   const function_application_exprt &f)
 {
   string_exprt str=get_string_expr(args(f, 3)[0]);
   const refined_string_typet &ref_type=to_refined_string_type(str.type());
-  const exprt &old_char=args(f, 3)[1];
-  const exprt &new_char=args(f, 3)[2];
-  string_exprt res=fresh_string(ref_type);
-
-  // We add axioms:
-  // a1 : |res| = |str|
-  // a2 : forall qvar, 0<=qvar<|res|,
-  //    str[qvar]=oldChar => res[qvar]=newChar
-  //    !str[qvar]=oldChar => res[qvar]=str[qvar]
-
-  m_axioms.push_back(res.axiom_for_has_same_length_as(str));
-
-  symbol_exprt qvar=fresh_univ_index("QA_replace", ref_type.get_index_type());
-  implies_exprt case1(
-    equal_exprt(str[qvar], old_char),
-    equal_exprt(res[qvar], new_char));
-  implies_exprt case2(
-    not_exprt(equal_exprt(str[qvar], old_char)),
-    equal_exprt(res[qvar], str[qvar]));
-  string_constraintt a2(qvar, res.length(), and_exprt(case1, case2));
-  m_axioms.push_back(a2);
-  return res;
+  if (const auto maybe_chars=try_char_char(args(f, 3)[1], args(f, 3)[2])) {
+    const auto old_char=maybe_chars->first;
+    const auto new_char=maybe_chars->second;
+    string_exprt res=fresh_string(ref_type);
+
+    // We add axioms:
+    // a1 : |res| = |str|
+    // a2 : forall qvar, 0<=qvar<|res|,
+    //    str[qvar]=oldChar => res[qvar]=newChar
+    //    !str[qvar]=oldChar => res[qvar]=str[qvar]
+
+    m_axioms.push_back(res.axiom_for_has_same_length_as(str));
+
+    symbol_exprt qvar=fresh_univ_index("QA_replace", ref_type.get_index_type());
+    implies_exprt case1(
+      equal_exprt(str[qvar], old_char),
+      equal_exprt(res[qvar], new_char));
+    implies_exprt case2(
+      not_exprt(equal_exprt(str[qvar], old_char)),
+      equal_exprt(res[qvar], str[qvar]));
+    string_constraintt a2(qvar, res.length(), and_exprt(case1, case2));
+    m_axioms.push_back(a2);
+    return res;
+  }
+  UNIMPLEMENTED;
```

Comment:
You could return `str` here instead of terminating the program. This would just generate a new string without any constraint.